aptos-move/mvhashmap/src/versioned_aggregators.rs

```diff
@@ -423,50 +425,197 @@ impl VersionedAggregators {
+                        ))
+                    } else {
+                        return Err(CommitError::CodeInvariantError(
+                            "Cannot apply delta to non-AggregatorValue::Aggregator".to_string(),
+                        ));
+                    }
+                },
+                AggregatorEntry::Apply(SnapshotDelta {
+                    base_aggregator,
+                    delta,
+                }) => {
+                    let prev_value = self.values
+                        .get_mut(base_aggregator)
+                        .ok_or_else(|| CommitError::CodeInvariantError("Cannot find base_aggregator for Apply(SnapshotDelta) during commit".to_string()))?
+                        .read_latest_committed_value(idx_to_commit)
+                        .map_err(|e| CommitError::CodeInvariantError(format!("Cannot read latest committed value for base aggregator for ApplySnapshotDelta) during commit: {:?}", e)))?;
+
+                    if let AggregatorValue::Aggregator(base) = prev_value {
+                        let new_value = delta.apply_to(base).map_err(|e| {
+                            CommitError::ReExecutionNeeded(format!(
+                                "Failed to apply delta to base: {:?}",
+                                e
+                            ))
+                        })?;
+                        Some(AggregatorEntry::Value(
+                            AggregatorValue::Snapshot(new_value),
+                            Some(SnapshotDelta {
+                                base_aggregator: *base_aggregator,
+                                delta: *delta,
+                            }),
+                        ))
+                    } else {
+                        return Err(CommitError::CodeInvariantError(
+                            "Cannot apply delta to non-AggregatorValue::Aggregator".to_string(),
+                        ));
+                    }
+                },
+                AggregatorEntry::Apply(SnapshotDerived { .. }) => {
+                    // Because Derived values can depend on the current value, we need to compute other values before it.
+                    derived_ids.push(id);
+                    None
+                },
+                AggregatorEntry::Estimate(_) => {
+                    return Err(CommitError::CodeInvariantError(
+                        "Cannot commit an estimate".to_string(),
+                    ))
+                },
+            };
+
+            if let Some(new_entry) = new_entry {
+                versioned_value.insert(idx_to_commit, new_entry);
+            }
+        }
+
+        for id in derived_ids {
+            let mut versioned_value = self
+                .values
+                .get_mut(&id)
+                .expect("Value in commit needs to be in the HashMap");
+            let entry_to_commit = versioned_value
+                .versioned_map
+                .get(&idx_to_commit)
```

Comment:
we should be able to remove here, and avoid formula clone later?
edit: we can remove above (in the first loop), then put the entry in derived_entries (instead of derived_ids), and iterate over them here, that way we also avoid code duplication and re-locking.

Reply:
good point, I didn't think, as usually we touch just one path and hold the lock.